Warp IO workload errors out with "The specified key does not exist" on Bare Metal

### Environment info
- NooBaa Version: 
```
[root@hpo-app11 ~]# noobaa version
INFO[0000] CLI version: 5.10.1
INFO[0000] noobaa-image: noobaa/noobaa-core:5.10.1
INFO[0000] operator-image: noobaa/noobaa-operator:5.10.1
[root@hpo-app11 ~]#
```

- Platform: 

```
[root@hpo-app11 ~]# oc version
Client Version: 4.11.0
Kustomize Version: v4.5.4
Server Version: 4.11.0
Kubernetes Version: v1.24.0+9546431
[root@hpo-app11 ~]#
```

ODF Version

```
[root@hpo-app11 ~]# oc get csv
NAME                                   DISPLAY                       VERSION               REPLACES                          PHASE
mcg-operator.v4.11.1                   NooBaa Operator               4.11.1                mcg-operator.v4.11.0              Succeeded
metallb-operator.4.11.0-202208300306   MetalLB Operator              4.11.0-202208300306                                     Succeeded
ocs-operator.v4.11.1                   OpenShift Container Storage   4.11.1                ocs-operator.v4.11.0              Succeeded
odf-csi-addons-operator.v4.11.1        CSI Addons                    4.11.1                odf-csi-addons-operator.v4.11.0   Succeeded
odf-operator.v4.11.1                   OpenShift Data Foundation     4.11.1                odf-operator.v4.11.0              Succeeded
[root@hpo-app11 ~]#

```

Warp version:

```
[root@hpo-app11 ~]# warp --version
warp version 0.5.5 - 1baadbc
[root@hpo-app11 ~]#
```

### Actual behavior
warp run on Bare Metal fails on all three metallb Endpoints


### Expected behavior
warp should execute successfully

### Steps to reproduce
1. Setup BM with CNSA/CSI RC4, ODF 4.11.1, latest DAS operator. 
2. Run warp on all three metallb IPs with 3 different users:

warp mixed --host=10.49.0.109 --access-key=uIBpl5fBWanFrXxLY6JW --secret-key=Z658YPATWG491XEbnn9ZwCPUGUS8KS36m91zWhUO --obj.size=1K --duration=120m --bucket=bucket11 --debug --insecure --tls 1> capture1.txt 2> error1.txt

warp mixed --host=10.49.0.110 --access-key=lDW9JSyitHuYa0EGx6vM --secret-key=BAu2KPpbAct28rvYwCdjILkO1vNWphsO57H5htlD --obj.size=1K --duration=120m --bucket=bucket22 --debug --insecure --tls 1> capture2.txt 2> error2.txt

warp mixed --host=10.49.0.111 --access-key=8FnXjkMLesfOOOr4TKw6 --secret-key=OjcPkdOyTE6FDujcmBj3JeUVl11TpZznyj4GyVhe --obj.size=1M --duration=120m --bucket=bucket33 --debug --insecure --tls 1> capture3.txt 2> error3.txt




Attachments:
[capture2.txt](https://github.com/noobaa/noobaa-core/files/9599718/capture2.txt)
[capture3.txt](https://github.com/noobaa/noobaa-core/files/9599725/capture3.txt)
[capture1.txt](https://github.com/noobaa/noobaa-core/files/9599713/capture1.txt)
[error1.txt](https://github.com/noobaa/noobaa-core/files/9599747/error1.txt)
[error2.txt](https://github.com/noobaa/noobaa-core/files/9599748/error2.txt)
[error3.txt](https://github.com/noobaa/noobaa-core/files/9599749/error3.txt)


Logs: https://ibmandredhatguests.slack.com/archives/G015Z7SDWQ0/p1663598668494769?thread_ts=1663597420.548999&cid=G015Z7SDWQ0

@romayalon 
On a freshly deployed DAS cluster (ODF 4.11.1 GA) we again hit the duplicate ids in bucketstats table. 

```
Sep-26 6:32:07.896 [Endpoint/14] [ERROR] CONSOLE:: RPC._on_request: ERROR srv object_api.update_endpoint_stats reqid 19524@fcall://fcall(7om8vqvf) connid fcall://fcall(7om8vqvf) AssertionError [ERR_ASSERTION]: _id must be unique. found 2 rows with _id=undefined in table bucketstats
Sep-26 6:32:07.897 [Endpoint/14] [ERROR] core.rpc.rpc:: RPC._request: response ERROR srv object_api.update_endpoint_stats reqid 19524@fcall://fcall(7om8vqvf) connid fcall://fcall(7om8vqvf) params { namespace_stats: [ { io_stats: { read_count: 2199279, write_count: 929200, read_bytes: 55346668240896, write_bytes: 13374358598656, error_write_bytes: 0, error_write_count: 0, error_read_bytes: 0, error_read_count: 0 }, namespace_resource_id: '632d5b3674e74100298682d4' }, [length]: 1 ], bucket_counters: [ { bucket_name: SENSITIVE-d11ed9bf0f42c55a, content_type: 'application/octet-stream', read_count: 1055154, write_count: 358804 }, { bucket_name: SENSITIVE-40584c364915f5f3, content_type: 'application/octet-stream', read_count: 1144123, write_count: 374277 }, [length]: 2 ] } took [8.8+0.4=9.2] [RpcError: _id must be unique. found 2 rows with _id=undefined in table bucketstats] { rpc_code: 'INTERNAL', rpc_data: { retryable: true } }
Sep-26 6:32:07.897 [Endpoint/14] [ERROR] core.sdk.endpoint_stats_collector:: failed on update_endpoint_stats. trigger_send_stats again [RpcError: _id must be unique. found 2 rows with _id=undefined in table bucketstats] { rpc_code: 'INTERNAL', rpc_data: { retryable: true } }
Sep-26 6:32:37.907 [Endpoint/14] [ERROR] core.util.postgres_client:: updateOneWithClient failed { system: 632d5af574e74100298682c0, bucket: 632f441da43595b2582184de, content_type: 'application/octet-stream' } { '$set': { last_write: 1664173957897, last_read: 1664173957897, system: 632d5af574e74100298682c0, bucket: 632f441da43595b2582184de, content_type: 'application/octet-stream' }, '$inc': { writes: 358804, reads: 1055154 } } UPDATE bucketstats SET data = jsonb_set(jsonb_set(jsonb_set(jsonb_set(jsonb_set(jsonb_set(jsonb_set(data,'{content_type}','"application/octet-stream"'),'{bucket}','"632f441da43595b2582184de"'),'{system}','"632d5af574e74100298682c0"'),'{last_read}','1664173957897'::jsonb),'{last_write}','1664173957897'::jsonb),'{reads}',to_jsonb(COALESCE(Cast(data->>'reads' as numeric),0)+1055154)),'{writes}',to_jsonb(COALESCE(Cast(data->>'writes' as numeric),0)+358804)) WHERE (data->>'system'='632d5af574e74100298682c0' and data->>'bucket'='632f441da43595b2582184de' and data->>'content_type'='application/octet-stream') RETURNING _id, data AssertionError [ERR_ASSERTION]: _id must be unique. found 2 rows with _id=undefined in table bucketstats


```

This issue re-occurred after our Thus discussion. Unfortunately as two people are working on the system, one was cleaning the DAS and eventually the logs are gone before the "noobaa diagnostics" was hit. Bottom line this is an issue in the latest code level of OCP 4.11 with ODF 4.11.1 

We will keep an eye for future runs and collect the logs. 

Need the workaround steps for documentation

--- a/src/util/postgres_client.js
+++ b/src/util/postgres_client.js
@@ -1207,17 +1207,23 @@ class PostgresTable {
         // eslint-disable-next-line no-constant-condition
         while (true) {
             let pg_client;
+            let locked;
             try {
                 pg_client = await this.client.pool.connect();
                 let update_res = await this._updateOneWithClient(pg_client, query, update, options);
                 if (update_res.rowCount === 0) {
-                    // try to lock the advisory_lock_key for this table and insert the first doc
-                    let locked = await this.try_lock_table(pg_client);
+                    // try to lock the advisory_lock_key for this table, try update and insert the first doc if 0 docs updated
+                    locked = await this.try_lock_table(pg_client);
                     if (locked) {
-                        const data = { _id: this.client.generate_id() };
-                        await this.insertOne(data);
-                        update_res = await this.updateOne(data, update, options);
+                        // try update again to avoid race conditions
+                        update_res = await this._updateOneWithClient(pg_client, query, update, options);
+                        if (update_res.rowCount === 0) {
+                            const data = { _id: this.client.generate_id() };
+                            await this.insertOne(data);
+                            update_res = await this.updateOne(data, update, options);
+                        }
                         await this.unlock_table(pg_client);
+                        locked = false;
                     } else {
                         // lock was already taken which means that another call to findOneAndUpdate should have inserted.
                         // throw and retry
@@ -1238,7 +1244,11 @@ class PostgresTable {
                     throw err;
                 }
             } finally {
-                pg_client.release();
+                if (locked) {
+                    await this.unlock_table(pg_client);
+                    locked = false;
+                }
+                if (pg_client) pg_client.release();
             }
         }
     }

